README.md
# doilinker

<!-- badges: start -->
[![R-CMD-check](https://github.com/lazappi/doilinker/workflows/R-CMD-check/badge.svg)](https://github.com/lazappi/doilinker/actions)
[![Codecov test coverage](https://codecov.io/gh/lazappi/doilinker/branch/main/graph/badge.svg)](https://codecov.io/gh/lazappi/doilinker?branch=main)
<!-- badges: end -->

**{doilinker}** allows you to link related DOIs.
Usually this is to find the published version of a preprint (or the reverse).
It does this by querying the Crossref database using **{rcrossref}** and matching DOIs using the method described by Cabanac, Oikonomidi and Boutron (see citation below).

Documentation for the package can be viewed at http://lazappi.github.io/doilinker.

## Installation

The package can be installed from GitHub:

```r
# install.packages("remotes")
remotes::install_github("lazappi/doilinker")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(doilinker)

search_doi_links("10.1101/133173")
```

## Code of Conduct
  
Please note that the doilinker project is released with a [Contributor Code of Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html). By contributing to this project, you agree to abide by its terms.

## Citation

If you use this package please cite the original publication for the linking method
(if you want to cite the package directly as well that would be great 😉 but crediting the original authors is the important part):

> Cabanac G, Oikonomidi T, Boutron I. "Day-to-day discovery of preprint-publication links." **Scientometrics**. 2021;1–20. DOI: [10.1007/s11192-021-03900-7](https://doi.org/10.1007/s11192-021-03900-7)
>

```bibtex
@ARTICLE{Cabanac2021-ge,
  title    = "Day-to-day discovery of preprint-publication links",
  author   = "Cabanac, Guillaume and Oikonomidi, Theodora and Boutron, Isabelle",
  journal  = "Scientometrics",
  pages    = "1--20",
  month    =  apr,
  year     =  2021,
  language = "en"
}
```


NEWS.md
# doilinker 0.1.0 (2021-06-25)

* First version with basic functionality


tests/testthat.R
library(testthat)
library(doilinker)

test_check("doilinker")


tests/testthat/test-search.R
test_that("basic search works", {
    results <- search_doi_links("10.1101/133173")
    expect_s3_class(results, "tbl_df")
    expect_true(all(c("doi", "score", "match") %in% colnames(results)))
})


R/checks.R
#' Is DOI preprint
#'
#' Check whether a DOI is from a preprint server.
#'
#' @param doi DOI to check
#' @param container_title Whether or not there is a container title (journal
#' name) associated with the DOI
#'
#' @return Logical whether the DOI is from a preprint
is_doi_preprint <- function(doi, container_title) {

    peerj    <- stringr::str_detect(doi, "^10.7287/")
    square   <- stringr::str_detect(doi, "^10.21203/")
    biorxiv  <- stringr::str_detect(doi, "^10.1101/") && !container_title
    preprint <- biorxiv || peerj || square

    return(preprint)
}

#' Is result a match
#'
#' Decide if a results from a DOI query is a match or not.
#'
#' @param query_title Title of the query reference
#' @param result_title Title of the result reference
#' @param query_orcid ORCiD of the query first author
#' @param result_orcid ORCiD of the result first author
#' @param query_authors Authors string for the query
#' @param result_authors Authors string for the result
#'
#' @details
#' Result is a match if the similarity between titles is sufficiently high or
#' if the similarity is lower by the first author ORCiDs match or the first
#' author names match.
#'
#' @return Logical whether the result is a match or not
is_result_match <- function(query_title, result_title, query_orcid,
                            result_orcid, query_authors, result_authors) {
    title_similarity <- calc_jaccard_similarity(query_title, result_title)

    if (title_similarity >= 0.8) {
        # High title similarity
        return(TRUE)
    } else if (title_similarity >= 0.1) {
        # Low title similarity...
        if (!is.na(query_orcid) && !is.na(result_orcid) &&
            (result_orcid == query_orcid)) {
            # ...and same first author ORCiD
            return(TRUE)
        }
        if (same_first_author(query_authors, result_authors)) {
            # ...or same first author name
            return(TRUE)
        }
    }

    return(FALSE)
}


R/authors.R
#' Get Crossref authors string
#'
#' Convert the authors table from Crossref into a single string. It has the form
#' "LastName1, FirstName1; LastName2, FirstName2;...".
#'
#' @param cr_authors Authors data.frame from a Crossref query
#'
#' @return Character vector with authors names
get_cr_authors_str <- function(cr_authors) {

    has_family <- "family" %in% colnames(cr_authors)
    has_given  <- "given" %in% colnames(cr_authors)
    has_name   <- "name" %in% colnames(cr_authors)

    if (has_family) {
        if (has_given) {
            authors <- paste0(cr_authors$family, ", ", cr_authors$given)
        } else {
            authors <- cr_authors$family
        }
    } else if (has_name) {
        authors <- cr_authors$name
    } else {
        authors <- ""
    }

    authors_str <- paste(authors, collapse = "; ")

    return(authors_str)
}

#' Get first author ORCiD
#'
#' Return the ORCiD for the first author of a reference.
#'
#' @param cr_authors Authors data.frame from a Crossref query
#'
#' @return Character vector with ORCiD or `NA` if not found
get_first_orcid <- function(cr_authors) {

    if (!("ORCID" %in% colnames(cr_authors))) {
        return(NA)
    }

    orcid <- cr_authors$ORCID[1]

    return(orcid)
}

#' Same first author
#'
#' Check if two author strings have the same first author.
#'
#' @param authors_str1 Character vector containing first authors list
#' @param authors_str2 Character vector containing second authors list
#'
#' @details
#' Last names of the first authors are compared. If these match then first names
#' are checked and `TRUE` is returned if either the whole first name or first
#' initial matches. If first and last names cannot be separated the whole author
#' names are compared. Some simplification of characters in names is performed
#' to improve matches.
#'
#' @return Logical whether first authors are the same
same_first_author <- function(authors_str1, authors_str2) {

    authors_str1 <- stringi::stri_trans_general(authors_str1, "Latin-ASCII")
    authors_str2 <- stringi::stri_trans_general(authors_str2, "Latin-ASCII")

    author1 <- stringr::str_split(authors_str1, ";")[[1]][1]
    author2 <- stringr::str_split(authors_str2, ";")[[1]][1]

    author1 <- stringr::str_to_lower(stringr::str_replace(author1, "-", " "))
    author2 <- stringr::str_to_lower(stringr::str_replace(author2, "-", " "))

    author_names1 <- stringr::str_squish(stringr::str_split(author1, ",")[[1]])
    author_names2 <- stringr::str_squish(stringr::str_split(author2, ",")[[1]])

    # No separable first/last names
    if (length(author_names1) == 1 || length(author_names1) == 2) {
        return(author1 == author2)
    }

    # Last names don't match
    if (author_names1[1] != author_names2[1]) {
        return(FALSE)
    }

    # First names match
    if (author_names1[2] == author_names2[2]) {
        return(TRUE)
    }

    initial1 <- stringr::str_sub(author_names1[2], start = 1, end = 1)
    initial2 <- stringr::str_sub(author_names2[2], start = 1, end = 1)

    # First initials match
    if (initial1 == initial2) {
        return(TRUE)
    }

    return(FALSE)
}


R/similarity.R
#' Calculate Jaccard similarity
#'
#' Calculate the Jaccard similarity between two strings. Strings are first
#' tokenised, stop words are removed and tokens are stemmed.
#'
#' @param string1 First string to score similarity
#' @param string2 Second string to score similarity
#'
#' @return Jaccard similarity score
calc_jaccard_similarity <- function(string1, string2) {

    string1 <- stringi::stri_trans_general(string1, "Latin-ASCII")
    string2 <- stringi::stri_trans_general(string2, "Latin-ASCII")

    tokens1 <- stringr::str_split(
        stringr::str_to_lower(string1), "[^\\w\\d/-]"
    )[[1]]
    tokens2 <- stringr::str_split(
        stringr::str_to_lower(string2), "[^\\w\\d/-]"
    )[[1]]

    stopwords <- c(corpus::stopwords_en, "")
    tokens1 <- corpus::stem_snowball(tokens1[!(tokens1 %in% stopwords)])
    tokens2 <- corpus::stem_snowball(tokens2[!(tokens2 %in% stopwords)])

    jaccard <- length(intersect(tokens1, tokens2)) /
        length(union(tokens1, tokens2))

    return(jaccard)
}


R/search.R
#' Search DOI links
#'
#' Search Crossref for potential linked references for a DOI
#'
#' @param doi DOI to query
#' @param preprint Logical. Whether `doi` is a preprint (`TRUE`) or publication
#' (`FALSE`). If `NULL` (default) attempt to determine this from `doi`.
#' @param limit Number of Crossref results to return. Default is 20, max is
#' 1000.
#' @param filter_matches Logical. If `TRUE` only return matches rather than all
#' Crossref results.
#' @param verbose Logical. Whether or not to print progress messages.
#'
#' @details
#' Based on the method described in "Day-to-day discovery of
#' preprint–publication links" https://doi.org/10.1007/s11192-021-03900-7 and
#' code from https://github.com/gcabanac/preprint-publication-linker. Query
#' filters have been modified to allow reverse linking publications to preprints
#' as well as preprints to publications.
#'
#' @return tibble of potential links
#' @export
#'
#' @examples
#' search_doi_links("10.1101/133173")
search_doi_links <- function(doi, preprint = NULL, limit = 20,
                             filter_matches = FALSE, verbose = TRUE) {

    if (verbose) {message("Retrieving query data from Crossref...")}
    query_data <- rcrossref::cr_works(doi)$data

    query_title   <- query_data$title
    query_authors <- get_cr_authors_str(query_data$author[[1]])
    query_issued  <- query_data$issued
    query_orcid   <- get_first_orcid(query_data$author[[1]])

    if (is.null(preprint)) {
        query_preprint <- is_doi_preprint(
            doi,
            container_title = "container.title" %in% colnames(query_data)
        )
        if (verbose) {
            if (query_preprint) {
                message("Decided query is a preprint")
            } else {
                message("Decided query is a publication")
            }
        }
    } else {
        query_preprint <- preprint
    }

    if (query_preprint) {
        filters <- c(
            from_created_date = query_issued,
            type              = "journal-article",
            type              = "proceedings-article",
            type              = "book-chapter",
            type              = "book-part",
            type              = "book-section"
        )
    } else {
        filters <- c(
            until_created_date = query_issued,
            type               = "posted-content"
        )
    }

    if (verbose) {message("Retrieving results data from Crossref...")}
    results <- rcrossref::cr_works(
        query = "", # Fails if query is not set
        flq   = c(
            query.bibliographic = query_title,
            query.author        = query_authors
        ),
        filter = filters,
        limit  = limit,
        sort   = "score",
        order  = "desc"
    )$data
    if (verbose) {message("Found ", nrow(results), " results")}

    if (verbose) {message("Checking results for matches...")}
    is_match <- purrr::map_lgl(seq_len(nrow(results)), function(.idx) {

        results_row <- results[.idx, ]

        result_title   <- results_row$title
        result_authors <- get_cr_authors_str(results_row$author[[1]])
        result_orcid   <- get_first_orcid(results_row$author[[1]])

        is_result_match(query_title, result_title, query_orcid, result_orcid,
                        query_authors, result_authors)
    })

    results$match <- is_match

    if (filter_matches) {
        if (verbose) {message("Filtering matches...")}
        results <- results[is_match, ]
    }

    if (verbose) {message("Done!")}
    return(results)
}
